File: Java/001_two_sum.java
/*
Problem: 1 - Two Sum
Difficulty: Easy
Link: https://leetcode.com/problems/two-sum/

Problem Statement:
Given an array of integers nums and an integer target, return indices of the 
two numbers such that they add up to target.

Approach:
Use HashMap to store complement values and their indices.
For each element, check if (target - current) exists in the map.

Time Complexity: O(N)
Space Complexity: O(N)
*/
import java.util.*;

class Solution {
    public int[] twoSum(int[] nums, int target) {
        // HashMap to store {value: index}
        Map<Integer, Integer> seen = new HashMap<>();
        
        // Iterate through array
        for (int i = 0; i < nums.length; i++) {
            // Calculate complement
            int complement = target - nums[i];
            
            // Check if complement exists
            if (seen.containsKey(complement)) {
                return new int[] {seen.get(complement), i};
            }
            
            // Store current number
            seen.put(nums[i], i);
        }
        
        return new int[] {};
    }
    
    // Test cases
    public static void main(String[] args) {
        Solution sol = new Solution();
        
        // Test 1
        int[] nums1 = {2, 7, 11, 15};
        System.out.println("Output: " + Arrays.toString(sol.twoSum(nums1, 9)));
        // Expected: [0, 1]
        
        // Test 2
        int[] nums2 = {3, 2, 4};
        System.out.println("Output: " + Arrays.toString(sol.twoSum(nums2, 6)));
        // Expected: [1, 2]
    }
}

File: Cpp/001_two_sum.cpp
/*
Problem: 1 - Two Sum
Difficulty: Easy
Link: https://leetcode.com/problems/two-sum/

Approach:
Use unordered_map to store visited elements and their indices.
Time: O(N), Space: O(N)
*/
#include <vector>
#include <unordered_map>
#include <iostream>
using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        // Hash map: {value -> index}
        unordered_map<int, int> seen;
        
        for (int i = 0; i < nums.size(); i++) {
            int complement = target - nums[i];
            
            // Check if complement exists
            if (seen.find(complement) != seen.end()) {
                return {seen[complement], i};
            }
            
            // Store current element
            seen[nums[i]] = i;
        }
        
        return {};
    }
};

// Test cases
int main() {
    Solution sol;
    
    vector<int> nums1 = {2, 7, 11, 15};
    vector<int> result1 = sol.twoSum(nums1, 9);
    cout << "[" << result1[0] << ", " << result1[1] << "]" << endl;
    vector<int> nums2 = {3, 2, 4};
    vector<int> result2 = sol.twoSum(nums2, 6);
    cout << "[" << result2[0] << ", " << result2[1] << "]" << endl;

    return 0;
}